wagtail/images/views/images.py

```diff
@@ -359,10 +371,33 @@ def get(self, request, image_id, filter_spec):
         # Generate preview url
         preview_url = reverse("wagtailimages:preview", args=(image_id, filter_spec))
 
-        return JsonResponse(
-            {"url": site_root_url + url, "preview_url": preview_url}, status=200
+        message_labels = json.dumps(
+            {"400": self.invalid_filter_error},
+            cls=DjangoJSONEncoder,
         )
 
+        context["form"] = self.form
+        context["message_labels"] = message_labels
+        context["preview_url"] = preview_url
+        context["result_url"] = site_root_url + url
+
+        return context
+
+    def get_filter_spec(self, filter_method, width, height, closeness):
+        filter_spec = filter_method  # Default to 'original'
+
+        if filter_method == "width":
+            filter_spec += f"-{width}"
+        elif filter_method == "height":
+            filter_spec += f"-{height}"
+        elif filter_method in ["min", "max", "fill"]:
+            filter_spec += f"-{width}x{height}"
+            # Default closeness is 0 - avoid adding if not needed
+            if filter_method == "fill" and closeness != "0":
+                filter_spec += f"-c{closeness}"
+
+        return filter_spec
```

Comment:
Would be a nice use case for [pattern matching syntax](https://peps.python.org/pep-0622/), which we can use now that our support for Python 3.9 has ended!

Reply:
That's much cleaner. Love it.